plan/decorrelate.go

```diff
@@ -0,0 +1,82 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plan
+
+import (
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/util/types"
+)
+
+// extractCorColumnsBySchema only extracts the correlated columns that match the outer plan's schema.
+func extractCorColumnsBySchema(schema expression.Schema, innerPlan LogicalPlan) []*expression.CorrelatedColumn {
```

Comment:
If this function is only used for Apply, why not make it a method of Apply?
By the function name, we can't see where the argument `schema` and `innerPlan` came from.

And if you add an example in the comment, this would be much easier to understand.

Like given `t1`, `t2`,`t3`, `t1` is the outer table.
If schema is (`t2.c`), inner plan's schema is (`t1.c`, `t2.c`, `t3.c`), only `t2.c` is returned.